File: backend/src/main/java/com/innovategov/backend/service/RecommendationCalculator.java
package com.innovategov.backend.service;

import java.math.BigDecimal;
import java.math.RoundingMode;
import java.util.List;
import java.util.Locale;
import java.util.Set;

/**
 * Pure, dependency-free scoring logic for the Scale/Modify/Reject recommendation
 * engine (Section 7). Kept free of JPA/Spring so the decision math is directly
 * unit-testable with fixed inputs (Section 20).
 *
 * Design notes (documented since the spec doesn't fully pin these down):
 *  - performance_score: average, across KPIs with both a target and a recorded
 *    result, of how close the pilot came to target (capped at 1.0 per KPI so
 *    wildly overshooting one KPI can't mask failing others).
 *  - impact_score: breadth — the fraction of those KPIs that met or exceeded
 *    target. performance_score captures magnitude, impact_score captures breadth.
 *  - cost_score: this prototype has no per-pilot actual-vs-budgeted cost ledger,
 *    so cost efficiency is approximated by schedule adherence — the fraction of
 *    milestones completed without being marked DELAYED. A pilot that blew its
 *    timeline is treated as having blown its cost efficiency too, which is a
 *    reasonable proxy and, importantly, a transparent and reproducible one.
 *  - Some KPIs are "lower is better" (time-to-detect, false-positive rate, cost
 *    reduction targets phrased as a ceiling). Since the schema has no explicit
 *    direction column, direction is inferred from the KPI name via a keyword
 *    list — documented here rather than hidden.
 */
public final class RecommendationCalculator {

    // "reduction"/"increase"/"gain"/"improvement" KPIs are excluded even if they also match a
    // keyword below (e.g. "Survey Cost Reduction", "Loan Default Reduction", "Referral Time
    // Reduction") because the *recorded value already is the improvement/percentage achieved* —
    // higher is always better for those, regardless of the underlying quantity being named after
    // something you'd normally want to minimize (cost, time, default rate).
    private static final Set<String> HIGHER_IS_BETTER_OVERRIDE_KEYWORDS = Set.of(
            "reduction", "increase", "gain", "improvement"
    );
    private static final Set<String> LOWER_IS_BETTER_KEYWORDS = Set.of(
            "time", "delay", "latency", "rate"
    );

    public static final BigDecimal SCALE_THRESHOLD = new BigDecimal("0.70");
    public static final BigDecimal MODIFY_THRESHOLD = new BigDecimal("0.40");

    private RecommendationCalculator() {}

    public record KpiAssessment(String kpiName, BigDecimal targetValue, BigDecimal recordedValue) {}

    public record MilestoneAssessment(boolean delayed) {}

    public record ScoreResult(BigDecimal costScore, BigDecimal performanceScore, BigDecimal impactScore, BigDecimal overall, String recommendation) {}

    public static boolean isLowerIsBetter(String kpiName) {
        String normalized = kpiName.toLowerCase(Locale.ROOT);
        if (HIGHER_IS_BETTER_OVERRIDE_KEYWORDS.stream().anyMatch(normalized::contains)) {
            return false;
        }
        return LOWER_IS_BETTER_KEYWORDS.stream().anyMatch(normalized::contains);
    }

    /** Per-KPI achievement ratio in [0, 1] — 1.0 means target met or exceeded. */
    public static BigDecimal achievementRatio(KpiAssessment kpi) {
        if (kpi.targetValue() == null || kpi.targetValue().compareTo(BigDecimal.ZERO) == 0 || kpi.recordedValue() == null) {
            return null;
        }
        BigDecimal ratio;
        if (isLowerIsBetter(kpi.kpiName())) {
            if (kpi.recordedValue().compareTo(BigDecimal.ZERO) <= 0) {
                ratio = BigDecimal.ONE; // zero incidents/zero time is the best possible outcome
            } else {
                ratio = kpi.targetValue().divide(kpi.recordedValue(), 6, RoundingMode.HALF_UP);
            }
        } else {
            ratio = kpi.recordedValue().divide(kpi.targetValue(), 6, RoundingMode.HALF_UP);
        }
        return ratio.min(BigDecimal.ONE).max(BigDecimal.ZERO);
    }

    public static BigDecimal performanceScore(List<KpiAssessment> kpis) {
        List<BigDecimal> ratios = kpis.stream().map(RecommendationCalculator::achievementRatio)
                .filter(r -> r != null).toList();
        if (ratios.isEmpty()) return BigDecimal.ZERO;
        BigDecimal sum = ratios.stream().reduce(BigDecimal.ZERO, BigDecimal::add);
        return sum.divide(BigDecimal.valueOf(ratios.size()), 4, RoundingMode.HALF_UP);
    }

    public static BigDecimal impactScore(List<KpiAssessment> kpis) {
        List<BigDecimal> ratios = kpis.stream().map(RecommendationCalculator::achievementRatio)
                .filter(r -> r != null).toList();
        if (ratios.isEmpty()) return BigDecimal.ZERO;
        long metOrExceeded = ratios.stream().filter(r -> r.compareTo(BigDecimal.ONE) >= 0).count();
        return BigDecimal.valueOf(metOrExceeded).divide(BigDecimal.valueOf(ratios.size()), 4, RoundingMode.HALF_UP);
    }

    public static BigDecimal costScore(List<MilestoneAssessment> milestones) {
        if (milestones.isEmpty()) return BigDecimal.valueOf(0.5); // no schedule data — neutral, not penalized
        long onSchedule = milestones.stream().filter(m -> !m.delayed()).count();
        return BigDecimal.valueOf(onSchedule).divide(BigDecimal.valueOf(milestones.size()), 4, RoundingMode.HALF_UP);
    }

    public static ScoreResult compute(List<KpiAssessment> kpis, List<MilestoneAssessment> milestones) {
        BigDecimal cost = costScore(milestones);
        BigDecimal performance = performanceScore(kpis);
        BigDecimal impact = impactScore(kpis);
        BigDecimal overall = cost.add(performance).add(impact)
                .divide(BigDecimal.valueOf(3), 4, RoundingMode.HALF_UP);

        String recommendation;
        if (overall.compareTo(SCALE_THRESHOLD) >= 0) {
            recommendation = "SCALE";
        } else if (overall.compareTo(MODIFY_THRESHOLD) >= 0) {
            recommendation = "MODIFY";
        } else {
            recommendation = "REJECT";
        }
        return new ScoreResult(cost, performance, impact, overall, recommendation);
    }
}

File: backend/app/services/recommendation_calculator.py
"""
Scale / Modify / Reject scoring.

Direct port of Java's `RecommendationCalculator`. Pure and dependency-free, so
the decision maths can be tested against the Java implementation with fixed
inputs — which matters, because these three numbers are shown prominently on
the demo's recommendation screen.

The model, restated from the Java documentation:

``performance_score``
    Mean, across KPIs that have both a target and a recorded result, of how
    close the pilot came to target. Each KPI is capped at 1.0 so wildly
    overshooting one cannot mask failing others.
``impact_score``
    Breadth — the fraction of those KPIs that met or exceeded target.
    Performance captures magnitude; impact captures how many.
``cost_score``
    There is no per-pilot cost ledger in this schema, so cost efficiency is
    approximated by schedule adherence: the fraction of milestones not marked
    DELAYED. A pilot with no milestones scores a neutral 0.5 rather than 0.

``overall`` is the mean of the three, and the thresholds are 0.70 for SCALE and
0.40 for MODIFY.

**Direction.** Some KPIs are better when lower — time-to-detect, latency, a
false-positive rate. The schema has no direction column, so direction is
inferred from the KPI's name via two keyword sets, and the override set wins:
a "Survey Cost Reduction" KPI records *the improvement achieved*, so higher is
better even though it is named after a cost.

**`Decimal`, not float.** Java uses `BigDecimal` with explicit `HALF_UP`;
Python's default banker's rounding would disagree on exact halves.
"""
from __future__ import annotations

from dataclasses import dataclass
from decimal import ROUND_HALF_UP, Decimal

#: A KPI whose name contains one of these records an improvement, so higher is
#: always better — even when the underlying quantity (cost, time, a default
#: rate) is something you would normally want to minimise. Checked first.
HIGHER_IS_BETTER_OVERRIDE_KEYWORDS = frozenset({
    "reduction", "increase", "gain", "improvement",
})

#: Otherwise, a name containing one of these is treated as lower-is-better.
LOWER_IS_BETTER_KEYWORDS = frozenset({"time", "delay", "latency", "rate"})

SCALE_THRESHOLD = Decimal("0.70")
MODIFY_THRESHOLD = Decimal("0.40")

_RATIO_PRECISION = Decimal("0.000001")   # 6 dp, as `divide(..., 6, HALF_UP)`
_SCORE_PRECISION = Decimal("0.0001")     # 4 dp, as `divide(..., 4, HALF_UP)`

_ONE = Decimal("1")
_ZERO = Decimal("0")


@dataclass(frozen=True)
class KpiAssessment:
    kpi_name: str
    target_value: Decimal | None
    recorded_value: Decimal | None


@dataclass(frozen=True)
class MilestoneAssessment:
    delayed: bool


@dataclass(frozen=True)
class ScoreResult:
    cost_score: Decimal
    performance_score: Decimal
    impact_score: Decimal
    overall: Decimal
    recommendation: str


def is_lower_is_better(kpi_name: str) -> bool:
    """Infer KPI direction from its name. Overrides are checked first."""
    normalised = (kpi_name or "").lower()
    if any(keyword in normalised for keyword in HIGHER_IS_BETTER_OVERRIDE_KEYWORDS):
        return False
    return any(keyword in normalised for keyword in LOWER_IS_BETTER_KEYWORDS)


def achievement_ratio(kpi: KpiAssessment) -> Decimal | None:
    """
    Per-KPI achievement in [0, 1]; 1.0 means target met or exceeded.

    ``None`` when the KPI cannot be assessed — no target, a zero target, or no
    recorded result. Those are *excluded* from the averages rather than counted
    as zero, so an unmeasured KPI does not silently drag a pilot down.
    """
    if kpi.target_value is None or kpi.target_value == 0 or kpi.recorded_value is None:
        return None

    if is_lower_is_better(kpi.kpi_name):
        if kpi.recorded_value <= 0:
            # Zero incidents / zero elapsed time is the best possible outcome,
            # and avoids a division by zero.
            ratio = _ONE
        else:
            ratio = (kpi.target_value / kpi.recorded_value).quantize(
                _RATIO_PRECISION, rounding=ROUND_HALF_UP)
    else:
        ratio = (kpi.recorded_value / kpi.target_value).quantize(
            _RATIO_PRECISION, rounding=ROUND_HALF_UP)

    return max(_ZERO, min(_ONE, ratio))


def performance_score(kpis: list[KpiAssessment]) -> Decimal:
    """Mean achievement ratio across assessable KPIs. Magnitude."""
    ratios = [r for r in (achievement_ratio(k) for k in kpis) if r is not None]
    if not ratios:
        return _ZERO
    return (sum(ratios, _ZERO) / Decimal(len(ratios))).quantize(
        _SCORE_PRECISION, rounding=ROUND_HALF_UP)


def impact_score(kpis: list[KpiAssessment]) -> Decimal:
    """Fraction of assessable KPIs that met or exceeded target. Breadth."""
    ratios = [r for r in (achievement_ratio(k) for k in kpis) if r is not None]
    if not ratios:
        return _ZERO
    met = sum(1 for ratio in ratios if ratio >= _ONE)
    return (Decimal(met) / Decimal(len(ratios))).quantize(
        _SCORE_PRECISION, rounding=ROUND_HALF_UP)


def cost_score(milestones: list[MilestoneAssessment]) -> Decimal:
    """
    Schedule adherence, standing in for cost efficiency.

    No milestones means no schedule data, which is neutral (0.5) rather than a
    failing 0 — the pilot simply cannot be judged on this axis.
    """
    if not milestones:
        return Decimal("0.5")
    on_schedule = sum(1 for m in milestones if not m.delayed)
    return (Decimal(on_schedule) / Decimal(len(milestones))).quantize(
        _SCORE_PRECISION, rounding=ROUND_HALF_UP)


def compute(kpis: list[KpiAssessment],
            milestones: list[MilestoneAssessment]) -> ScoreResult:
    """The three component scores, their mean, and the resulting decision."""
    cost = cost_score(milestones)
    performance = performance_score(kpis)
    impact = impact_score(kpis)
    overall = ((cost + performance + impact) / Decimal(3)).quantize(
        _SCORE_PRECISION, rounding=ROUND_HALF_UP)

    if overall >= SCALE_THRESHOLD:
        recommendation = "SCALE"
    elif overall >= MODIFY_THRESHOLD:
        recommendation = "MODIFY"
    else:
        recommendation = "REJECT"

    return ScoreResult(cost_score=cost, performance_score=performance,
                       impact_score=impact, overall=overall,
                       recommendation=recommendation)


__all__ = [
    "HIGHER_IS_BETTER_OVERRIDE_KEYWORDS",
    "KpiAssessment",
    "LOWER_IS_BETTER_KEYWORDS",
    "MODIFY_THRESHOLD",
    "MilestoneAssessment",
    "SCALE_THRESHOLD",
    "ScoreResult",
    "achievement_ratio",
    "compute",
    "cost_score",
    "impact_score",
    "is_lower_is_better",
    "performance_score",
]
